sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/DeserializerBuildHelper.scala

```diff
@@ -161,6 +161,10 @@ object DeserializerBuildHelper {
     case _: StructType => expr
     case _: ArrayType => expr
     case _: MapType => expr
+    case _: DecimalType =>
+      // For Scala/Java `BigDecimal`, we accept decimal types of any valid precision/scale.
+      // Here we use the `DecimalType` object to indicate it.
+      UpCast(expr, DecimalType, walkedTypePath.getPaths)
```

Comment:
Can we just directly pass `expected` to `UpCast`, and don't change the definition of `UpCast`? I think that will fix it less invasively.

Reply:
hmm...I may not understand your concern here. If we do not change the  definition of `UpCast` from `DataType` to `AbstractDataType`, it can not accept **object** `DecimalType`.

Reply:
I mean `expected` here is a `DataType`. If we can't pass, how did the existing codes work before :-)..

Reply:
It's for `semantic` purpose. `UpCast` could still mean to cast a value to a concrete decimal type in this way while `object DecimalType` is an exception for the encoder.